src/i2c/src/i2c.rs

```diff
@@ -267,10 +289,10 @@ pub trait I2cDevice {
     fn funcs(&mut self) -> Result<u64>;
```

Comment:
in fact given we have an error exit case you could simplyfy it all to Ok(SmbusMsg) and drop the early return from the match expression and just have the result of the match expression as the return.

Reply:
Are you suggesting something like this ?

`diff --git a/src/i2c/src/i2c.rs b/src/i2c/src/i2c.rs
index 3561124e244f..b1a949bf95c3 100644
--- a/src/i2c/src/i2c.rs
+++ b/src/i2c/src/i2c.rs
@@ -177,44 +177,60 @@ impl SmbusMsg {
         let mut data = I2cSmbusData {
             block: [0; I2C_SMBUS_BLOCK_MAX + 2],
         };
-        let read_write: u8;
-        let mut smbus_data = None;
 
         // Write messages have only one request message, while read messages
         // will have two (except for few special cases of I2C_SMBUS_QUICK and
         // I2C_SMBUS_BYTE, where only one request messages is sent).
-        let size = match reqs.len() {
+        match reqs.len() {
             // Write requests (with some exceptions as mentioned above)
             1 => {
-                if (reqs[0].flags & I2C_M_RD) != 0 {
+                let read_write = if (reqs[0].flags & I2C_M_RD) != 0 {
                     // Special Read requests, reqs[0].len can be 0 or 1 only.
                     if reqs[0].len > 1 {
                         return Err(Error::MessageLengthInvalid("read", reqs[0].len as usize));
                     }
-                    read_write = I2C_SMBUS_READ;
+                    I2C_SMBUS_READ
                 } else {
-                    read_write = I2C_SMBUS_WRITE;
-                }
+                    I2C_SMBUS_WRITE
+                };
 
                 match reqs[0].len {
                     // Special Read requests
-                    0 => I2C_SMBUS_QUICK,
-                    1 => I2C_SMBUS_BYTE,
+                    0 => Ok(SmbusMsg {
+                        read_write,
+                        command: reqs[0].buf[0],
+                        size: I2C_SMBUS_QUICK,
+                        data: None,
+                    }),
+
+                    1 => Ok(SmbusMsg {
+                        read_write,
+                        command: reqs[0].buf[0],
+                        size: I2C_SMBUS_BYTE,
+                        data: None,
+                    }),
 
                     // Write requests
                     2 => {
                         data.byte = reqs[0].buf[1];
-                        smbus_data = Some(data);
-                        I2C_SMBUS_BYTE_DATA
+                        Ok(SmbusMsg {
+                            read_write,
+                            command: reqs[0].buf[0],
+                            size: I2C_SMBUS_BYTE_DATA,
+                            data: Some(data),
+                        })
                     }
+
                     3 => {
                         data.word = reqs[0].buf[1] as u16 | ((reqs[0].buf[2] as u16) << 8);
-                        smbus_data = Some(data);
-                        I2C_SMBUS_WORD_DATA
-                    }
-                    _ => {
-                        return Err(Error::MessageLengthInvalid("write", reqs[0].len as usize));
+                        Ok(SmbusMsg {
+                            read_write,
+                            command: reqs[0].buf[0],
+                            size: I2C_SMBUS_WORD_DATA,
+                            data: Some(data),
+                        })
                     }
+                    _ => Err(Error::MessageLengthInvalid("write", reqs[0].len as usize)),
                 }
             }
 
@@ -231,37 +247,31 @@ impl SmbusMsg {
                     || (reqs[0].len != 1)
                     || (reqs[1].len > 2)
                 {
-                    return Err(Error::SMBusTransferInvalid(
+                    Err(Error::SMBusTransferInvalid(
                         reqs.len(),
                         reqs[0].len,
                         reqs[1].len,
-                    ));
-                }
-                read_write = I2C_SMBUS_READ;
-                smbus_data = Some(data);
-
-                if reqs[1].len == 1 {
-                    I2C_SMBUS_BYTE_DATA
+                    ))
                 } else {
-                    I2C_SMBUS_WORD_DATA
+                    Ok(SmbusMsg {
+                        read_write: I2C_SMBUS_READ,
+                        command: reqs[0].buf[0],
+                        size: if reqs[1].len == 1 {
+                            I2C_SMBUS_BYTE_DATA
+                        } else {
+                            I2C_SMBUS_WORD_DATA
+                        },
+                        data: Some(data),
+                    })
                 }
             }
 
-            _ => {
-                return Err(Error::SMBusTransferInvalid(
-                    reqs.len(),
-                    reqs[0].len,
-                    reqs[1].len,
-                ));
-            }
-        };
-
-        Ok(SmbusMsg {
-            read_write,
-            command: reqs[0].buf[0],
-            size,
-            data: smbus_data,
-        })
+            _ => Err(Error::SMBusTransferInvalid(
+                reqs.len(),
+                reqs[0].len,
+                reqs[1].len,
+            )),
+        }
     }
 }
`

Reply:
Yes - basically having the nested matches all resolve to the either a Ok or Err instead of assembling something at the end. It would be nice to wrap return Err(Error::MessageLengthInvalid("read", reqs[0].len as usize)); into that but I can see it might be a special case.

From my (basic) rusty understanding having an exhaustive set of matches covering the whole state is preferable to having returns popping up in deeply nested if/else statements.

Reply:
Okay, I have pushed a separate patch for this now. I am not sure if it was better earlier or now :)

Lets see what you and @andreeaflorescu think about it.